pkg/config/validation/header_value_validator.go

```diff
@@ -0,0 +1,159 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package validation
+
+import (
+	"errors"
+	"fmt"
+)
+
+type ParserState int32
+
+const (
+	LiteralParserState                   ParserState = iota // processing literal data
+	VariableNameParserState                                 // consuming a %VAR% name
+	ExpectArrayParserState                                  // expect starting [ in %VAR([...])%
+	ExpectStringParserState                                 // expect starting " in array of strings
+	StringParserState                                       // consuming an array element string
+	ExpectArrayDelimiterOrEndParserState                    // expect array delimiter (,) or end of array (])
+	ExpectArgsEndParserState                                // expect closing ) in %VAR(...)%
+	ExpectVariableEndParserState                            // expect closing % in %VAR(...)%
+)
+
+// validateHeaderValue is golang port version of
+// https://github.com/envoyproxy/envoy/blob/master/source/common/router/header_parser.cc#L73
+func validateHeaderValue(headerValue string) error {
```

Comment:
can we add a fuzzer? Any parser should have fuzzing. See pkg/fuzz/README.md

Reply:
sure, will add fuzz test later.